### PR TITLE
WKBLoader

### LINKED ISSUE #767: WKBLoader
Well-known binary (WKB) is a binary geometry encoding. WKB encodes only geometries and doesn't store attributes. It's used in databases such as PostGIS and as the internal storage format of Shapefiles. It's also being discussed as the internal storage format for a "GeoArrow" specification. WKB is defined starting on page 62 of the [OGC Simple Features specification](http://portal.opengeospatial.org/files/?artifact_id=25355). 

It's essentially a binary representation of WKT. For common geospatial types including (Multi) `Point`, `Line`, and `Polygon`, there's a 1:1 correspondence between WKT/WKB and GeoJSON. WKT and WKB also support extended geometry types, such as `Curve`, `Surface`, and `TIN`, which don't have a correspondence to GeoJSON.

- Coordinates can be 2-4 dimensions and are interleaved.
- Positions stored as double precision

image

We currently use a fork of [`wellknown`](https://github.com/mapbox/wellknown) for WKT parsing, but `wellknown` only supports WKT and not WKB. 

The only permissive JS library I've seen to parse WKB is [`wkx`](https://github.com/cschwarz/wkx), which parses WKT and WKB (as well as EWKT and EWKB, which include spatial reference identifiers on each record). It uses Node buffers, which requires a polyfill in the browser. I [commented](https://github.com/cschwarz/wkx/issues/21#issuecomment-621472216) a month ago asking about interest in a PR to use `DataView`s instead of Node `Buffer`s, but haven't received a response from the maintainer yet. The project does also use an older prototype-based code style.

## Performance

There's momentum towards greater use of binary data transport in Vis.gl projects. There's potential for faster WKB parsing than existing alternatives by keeping decoded WKB geometries in typed arrays.

`wkx` appears to parse WKB to GeoJSON, which means there's presumably overhead both of creating many JS objects, and then also passing back to the main thread.

We couldn't avoid a pass over the WKB record entirely, but we could copy entire chunks of WKB data at a time into a `Float64Array` with minimal processing (possibly reversing byte endianness if needed, since WKB can be encoded in either big or little endianness).

### PR DESCRIPTION
**Pull request by kylebarron**
Closes #767

- How to handle NaNs in coordinates? Should they be skipped or included in `positions` as `NaN`?

Todo:

- [ ] binary to geojson converter, since the decoding creates TypedArrays of coordinates by default. Separate PR.
- [ ] preallocate array for entire geometry, instead of per linear ring, which then necessitates a copy when concatenating
- [ ] Spatial reference identifier (SRID) handling for Extended WKB. See #779.
- [ ] Option to create Float32Array for positions? Positions are always stored in WKB as doubles
- [ ] Support `GeometryCollection`s

[Other good visual WKB reference](https://docs.teradata.com/reader/bbnDyCxlS~Mtf8vF_Sz3mw/C8umIM0gcwJu6eobQtRRQQ)

### CODE REVIEW

File: modules/wkb/docs/api-reference/wkb-loader.md [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+<p class="badges">
+  <img src="https://img.shields.io/badge/From-v2.3-blue.svg?style=flat-square" alt="From-v2.3" />
+</p>
+
+Loader for the [Well-known binary] format for representation of geometry.
+
+| Loader                | Characteristic                                                                                               |
+| --------------------- | ------------------------------------------------------------------------------------------------------------ |
+| File Extension        | `.wkb`,                                                                                                      |
+| File Type             | Text                                                                                                         |
```
Comment (kylebarron) [inline modules/wkb/docs/api-reference/wkb-loader.md:12]: Update

File: modules/wkb/docs/api-reference/wkb-loader.md [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+```js
+import {WKBLoader} from '@loaders.gl/wkb';
+import {parseSync} from '@loaders.gl/core';
+
+const data = parseSync('LINESTRING (30 10, 10 30, 40 40)', WKTLoader);
+// => {type: 'LineString', coordinates: [[30, 10], [10, 30], [40, 40]]}
+```
+
+```js
+import {WKTLoader} from '@loaders.gl/wkt';
```
Comment (kylebarron) [inline modules/wkb/docs/api-reference/wkb-loader.md:30]: Update

File: modules/wkb/package.json [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+    "dist",
+    "README.md"
+  ],
+  "scripts": {
+    "pre-build": "npm run build-worker && npm run build-bundle && npm run build-bundle -- --env.dev",
+    "build-bundle": "webpack --display=minimal --config ../../scripts/bundle.config.js",
+    "build-worker": "webpack --entry ./src/wkt-loader.worker.js --output ./dist/wkt-loader.worker.js --config ../../scripts/worker-webpack-config.js"
+  },
+  "devDependencies": {
+    "fuzzer": "^0.2.1"
```
Comment (kylebarron) [inline modules/wkb/package.json:35]: Used with wkt, but can probably be removed here.

File: modules/wkb/src/lib/parse-wkb.js [hunk 4]
```diff
@@ -0,0 +1,220 @@
...
+// https://stackoverflow.com/a/55261098
+const cumulativeSum = sum => value => (sum += value);
+
+function parsePolygon(view, offset, dimension, littleEndian) {
+  const nRings = view.getUint32(offset, littleEndian);
+  offset += 4;
+
+  const rings = [];
+  for (let i = 0; i < nRings; i++) {
+    var {positions, offset} = parseLineString(view, offset, dimension, littleEndian);
```
Comment (kylebarron) [inline modules/wkb/src/lib/parse-wkb.js:92]: Probably best to not destructure, since I want to update `offset`, but declare `positions` as `const`

**Comment by kylebarron** [comment]
> * How to handle NaNs in coordinates? Should they be skipped or included in `positions` as `NaN`?

Now I believe all tests pass except missing data tests. I.e. some tests ported from `wkx` are, for example, a point of length 1 where both coordinates are float `NaN`s. It converts such a point to the WKT `POINT EMPTY`, and to GeoJSON as
```json
{"type": "Point", "coordinates": []}
```
I think my code currently would load it as:
```
{"type": "Point", "coordinates": [NaN, NaN]}
```
Should I silently skip over any coordinates that are `NaN`?

**Comment by ibgreen** [comment]
> It converts such a point to the WKT POINT EMPTY

Are empty points legal in WKT? That seems unnecessarily tricky to deal with in apps or on the GPU in particular, so I would prefer not returning empty points. 

> Should I silently skip over any coordinates that are NaN?

My initial thought would be just pass return them to the app. If this is common in actual files we may rethink it, but if this is just a corner case in the test suite it doesn't really matter.

Principle: Since we have multiple geospatial loaders, I think the answer should apply to all of them. I..e do the loaders do data cleaning, or does the app clean afterwards. We could always add an (optional?) data cleaning feature later.

Another option is to follow Arrow and have a nullArray (nullBitmap) i.e. we generate data for each point, but also an array of boolean values indicating which ones are valid. In which case it doesn't matter what values we store in that point.

**Comment by kylebarron** [comment]
> Are empty points legal in WKT? That seems unnecessarily tricky to deal with in apps or on the GPU in particular, so I would prefer not returning empty points.

I did a little research on this. Yes, all geometries can be `EMPTY` in WKT. So you can have `POINT EMPTY`, `LINESTRING EMPTY`, `POLYGON EMPTY`, `GEOMETRYCOLLECTION EMPTY` (I think it might also allow `MULTI* EMPTY`).

Paul Ramsey (core developer of PostGIS) [says](https://gis.stackexchange.com/a/106942)

> POINT EMPTY is not representable in WKB

by which I think he means the closest equivalent to `POINT EMPTY` in WKB is `Point(NaN NaN)`, since doubles can represent NaN.

> My initial thought would be just pass return them to the app. If this is common in actual files we may rethink it, but if this is just a corner case in the test suite it doesn't really matter.

I don't know how common it is, but given that NaN is a valid double, it could present itself in real WKB data.
 
> Principle: Since we have multiple geospatial loaders, I think the answer should apply to all of them. I..e do the loaders do data cleaning, or does the app clean afterwards. We could always add an (optional?) data cleaning feature later.

I agree. I think for now I'll just pass the NaN's through, and we can revisit this in the future.

**Comment by kylebarron** [comment]
Upon closer inspection, handling NaN values is only an issue for type `POINT EMPTY`. For LineStrings and Polygons, there's a `uint32` value that defines the number of doubles that follow, and that value can be 0. But for Points, there must always be two numbers (or more depending on dimension) that follow, so those are NaN.

**Comment by kylebarron** [comment]
It looks like this is failing because of Draco test errors.
